Домашнее задание 22 спринт

darkelf007 opened the pull request.

Commit:
Верстка New PlayList

Commit:
Добавлен New Playlist Database
Добавлен New Playlist Fragment
Архитектура New Playlist Fragment
Player Activity переделан в PlayerFragment
Исправлен цвет темного оформления
Настроено поведение ButtomNavigationView
Добавлена шторка для PlayerFragent
Исправлена ошибка поведения истории найденых треков
Настроены ссылки между фрагментами
fragment_newplaylist завернут в NestedScrollView для адекватного нажатия на кнопу "создать"

Commit:
Исправлена ошибка где track id превращался в NULL

sukhoikms27 commented on app/src/main/java/com/android/playlistmaker/main/ui/MainActivity.kt:
```diff
@@ -78,4 +113,8 @@ class MainActivity : AppCompatActivity() {
     }
+
+    private fun hideBottomNavigation(isHide: Boolean) {
+        bottomNavigationView.visibility = if (!isHide) View.VISIBLE else View.GONE
```
Рекоменадция:
У `view` есть удобный экстеншен: `View.isVisible = true/false`, которая заменяет привычное нам `View.visibility = View.VISIBLE/View.GONE`, ну и гораздо удобнее!
`bottomNavigationView.isVisible= !isHide`

sukhoikms27 commented on app/src/main/java/com/android/playlistmaker/media/presentation/PlaylistFragment.kt:
```diff
@@ -4,25 +4,150 @@ import android.os.Bundle
+        return current
+    }
+
+    private fun View.show(isVisible: Boolean) {
```
Рекомендация:
можно заменить на экстеншен `isVisible`

sukhoikms27 commented on app/src/main/java/com/android/playlistmaker/media/presentation/adapter/PlaylistAdapter.kt:
```diff
@@ -0,0 +1,82 @@
+        return file.toUri()
+    }
+
+    private fun pluralizeWord(number: Int, word: String): String {
```
Воспользуйся лучше plurals, в андроиде уже реализована логика, с помощью которой для каждой локали будет подбираться соответствующее окончание. Почитать о том, как это использовать, можно в официальной документации - https://developer.android.com/guide/topics/resources/string-resource#Plurals

sukhoikms27 commented on app/src/main/java/com/android/playlistmaker/new_playlist/presentation/NewPlaylistFragment.kt:
```diff
@@ -0,0 +1,217 @@
+
+        newPlayListButton.isEnabled = false
+
+        editNameEditText.addTextChangedListener(object : TextWatcher {
```
Рекомендация:
у поля ввода есть экстеншен `addTextChangedListener`, в котором есть лямбды, аналогичные методам в `textWatcher`, поэтому можно не создавать объект textWatcher, а пользоваться готовой лямбдой, так же, если тебе требуется только метод `onTextChanged()` - можно воспользоваться другим экстеншеном: `doOnTextChanged {}`

sukhoikms27 commented on app/src/main/java/com/android/playlistmaker/new_playlist/presentation/NewPlaylistFragment.kt:
```diff
@@ -0,0 +1,217 @@
+            )
+
+            viewLifecycleOwner.lifecycleScope.launch {
+                viewModel.insertPlaylistToDatabase(playlist)
```
Общее замечание к названиям интеракторов и методов на слоях ui/domain - в этих слоях не должны знать что мы используем в качестве хранилища, ведь мы так же можем хранить на сервере, в sharedPref'ах или еще где то, поэтому названия лучше сделать более лаконичными:
`addPlaylist/insertPlaylist`

sukhoikms27 commented on app/src/main/java/com/android/playlistmaker/new_playlist/presentation/NewPlaylistFragment.kt:
```diff
@@ -0,0 +1,217 @@
+            }.show()
+    }
+
+    private fun saveImageToPrivateStorage(uri: Uri, nameOfFile: String) {
```
это лучше обрабатывать на уровне вьюмодели или репозитория- работа с application context'ом там доступна

sukhoikms27 commented on app/src/main/java/com/android/playlistmaker/player/presentation/PlaylistBottomSheetAdapter.kt:
```diff
@@ -0,0 +1,78 @@
+        return file.toUri()
+    }
+
+    private fun pluralizeWord(number: Int, word: String): String {
```
Воспользуйся лучше plurals, в андроиде уже реализована логика, с помощью которой для каждой локали будет подбираться соответствующее окончание. Почитать о том, как это использовать, можно в официальной документации - https://developer.android.com/guide/topics/resources/string-resource#Plurals

sukhoikms27 commented on app/src/main/java/com/android/playlistmaker/player/presentation/PlaylistBottomSheetAdapter.kt:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    private fun getUriOfImageFromStorage(fileName: String): Uri {
+        val filePath = File(context.getExternalFilesDir(Environment.DIRECTORY_PICTURES), "myalbum")
```
адаптер не должен идти в файловую систему, данные для него должны быть готовы заранее (на уровне вьюмодели)

на слабых девайсах это может сильно влиять на производительность

аналогичное замечание для адаптера фрагмента плейлистов

Commit:
Исправлен crit. нельзя создать плейлист с пустым названием
Исправлен паттерн MVVM new_playlist
Добавлен update в DAO для проверки на дубликат id треков в плейлисте
В UI исправлено отображение списка плейлистов

Commit:
Исправление задержки кнопки плей

Commit:
перемещена обработка imageUri из адаптеров в репозиторий
обновлен PlayList для imageUri

Commit:
рефакторинг по рекомендации ревьюера

sukhoikms27 commented on app/src/main/java/com/android/playlistmaker/main/ui/MainActivity.kt:
```diff
@@ -75,46 +71,35 @@ class MainActivity : AppCompatActivity(), BottomNavigationListener {
             if (shouldHideBottomNav) {
-                if (bottomNavigationView.visibility != View.GONE) {
-                    bottomNavigationView.visibility = View.GONE
+                if (binding.bottomNavigationView.isVisible) {
```
условие лишнее - если `binding.bottomNavigationView` в состоянии gone, то строка в словии просто не выполнится (оптимизации)

sukhoikms27 replied:
и достаточно будет сделать:
`binding.bottomNavigationView.isVisible = !shouldHideBottomNav`

sukhoikms27 commented on app/src/main/java/com/android/playlistmaker/media/data/repository/PlaylistMediaDatabaseRepositoryImpl.kt:
```diff
@@ -1,19 +1,38 @@
     }
 
-
+    private fun getUriOfImageFromStorage(fileName: String): Uri? {
```
можно параметр сделать nullable и первой строкой проверять на Null и возвращать сразу null:
```
private fun getUriOfImageFromStorage(fileName: String?): Uri? {
    fileName ?: return null
    ...
}
```

тогда и метод будет более универсальным, и в методе выше позволит убрать не самый удобный if:
`val imageUri = getUriOfImageFromStorage(entity.filePath)`

sukhoikms27 commented on app/src/main/java/com/android/playlistmaker/new_playlist/data/Impl/FileRepositoryImpl.kt:
```diff
@@ -0,0 +1,39 @@
+class FileRepositoryImpl(
+    private val context: Context
+) : FileRepository {
+    override suspend fun saveImageToPrivateStorage(uri: Uri, nameOfFile: String): Result<Unit> {
```
если вся идея с Result чтобы поймать ошибку и не упасть с ней, то тогда лучше воспользоваться `runCatching`:
```
runCatching {
    ...
    выполняем все что обернули в try
    ...
}.onFailure {
    Log.e("ooops, fail", it)
}
```
и возвращать что либо в этом методе нет необходимости

sukhoikms27 reviewed (commented):
Все супер! Добавил пару рекомендаций, в остальном все ок!

Commit:
доп. рефакторинг по рекомендации ревьюера
